File: next/Process/ProcessFlow.fs
namespace Wanxiangshu.Next.Process

open System
open Wanxiangshu.Next.Kernel

module ProcessFlows =

    type ProcessFlow<'a> = Flow<ProcessContext, ProcessError, 'a>

    let execute (cmd: Command) : ProcessFlow<Fact.ProcessResult> =
        Flow.create (fun ctx ct ->
            task {
                let effectiveCmd =
                    match cmd.Deadline, ctx.DefaultTimeout with
                    | Some _, _ -> cmd
                    | None, Some budget ->
                        { cmd with
                            Deadline = Some(Deadline.ofBudget DateTimeOffset.UtcNow budget) }
                    | None, None -> cmd

                let! spawnResult = ProcessSpawn.spawn effectiveCmd (Some ctx) ct

                match spawnResult with
                | Error e -> return Error e
                | Ok handle ->
                    use h = handle
                    return! Flow.run ctx ct (h.RunToCompletion())
            })

    let runFlow
        (ctx: ProcessContext)
        (ct: System.Threading.CancellationToken)
        (flow: ProcessFlow<'a>)
        : System.Threading.Tasks.Task<Result<'a, ProcessError>> =
        task {
            try
                return! Flow.run ctx ct flow
            with ex ->
                let msg = if isNull (box ex) then "" else string ex

                if msg.Contains("cancel") || msg.Contains("Cancel") || msg.Contains("Operation") then
                    return Error(ProcessError.ProcessCancelled "Operation cancelled")
                else
                    return Error(ProcessError.ExecutionFailed msg)
        }

File: .wanxiangshu/runtimes/Process/ProcessFlow.js

import { Flow_run, Flow_create } from "../Kernel/Flow.js";
import { task } from "../fable_modules/fable-library-js.5.6.0/TaskBuilder.js";
import { DeadlineModule_ofBudget } from "./Deadline.js";
import { utcNow } from "../fable_modules/fable-library-js.5.6.0/DateOffset.js";
import { Command } from "./Command.js";
import { ProcessSpawn_spawn } from "./ProcessHandle.js";
import { FSharpResult$2 } from "../fable_modules/fable-library-js.5.6.0/Result.js";
import { Operators_IsNull } from "../fable_modules/fable-library-js.5.6.0/FSharp.Core.js";
import { toString } from "../fable_modules/fable-library-js.5.6.0/Types.js";
import { ProcessError } from "./ProcessTypes.js";

export function execute(cmd) {
    return Flow_create((ctx, ct) => {
        const builder$0040 = task();
        return builder$0040.Run(builder$0040.Delay(() => {
            let effectiveCmd;
            const matchValue_1 = ctx.DefaultTimeout;
            if (cmd.Deadline == null) {
                if (matchValue_1 == null) {
                    effectiveCmd = cmd;
                }
                else {
                    const budget = matchValue_1;
                    effectiveCmd = (new Command(cmd.FileName, cmd.Arguments, cmd.WorkingDirectory, cmd.Environment, cmd.Stdin, DeadlineModule_ofBudget(utcNow(), budget), cmd.PtyOptions));
                }
            }
            else {
                effectiveCmd = cmd;
            }
            return builder$0040.Bind(ProcessSpawn_spawn(effectiveCmd, ctx, ct), (_arg) => {
                const spawnResult = _arg;
                return (spawnResult.tag === 0) ? builder$0040.Using(spawnResult.fields[0], (_arg_1) => builder$0040.ReturnFrom(Flow_run(ctx, ct, _arg_1.RunToCompletion()))) : builder$0040.Return(new FSharpResult$2(1, [spawnResult.fields[0]]));
            });
        }));
    });
}

export function runFlow(ctx, ct, flow) {
    const builder$0040 = task();
    return builder$0040.Run(builder$0040.Delay(() => builder$0040.TryWith(builder$0040.Delay(() => builder$0040.ReturnFrom(Flow_run(ctx, ct, flow))), (_arg) => {
        const ex = _arg;
        const msg = Operators_IsNull(ex) ? "" : toString(ex);
        return (((msg.indexOf("cancel") >= 0) ? true : (msg.indexOf("Cancel") >= 0)) ? true : (msg.indexOf("Operation") >= 0)) ? builder$0040.Return(new FSharpResult$2(1, [new ProcessError(1, ["Operation cancelled"])])) : builder$0040.Return(new FSharpResult$2(1, [new ProcessError(3, [msg])]));
    })));
}

